File: ObjectiveC/BeerSample/BeerSample/StoreViewController.m
//
//  StoreViewController.m
//  BeerSample
//
//  Created by Vinita Miranda on 3/6/15.
//  Copyright (c) 2015 Vinita Miranda. All rights reserved.
//

#import "StoreViewController.h"
#import "StoreTableViewCell.h"
#import "MapViewController.h"

@interface StoreViewController ()
@property(nonatomic,weak)IBOutlet UITableView *tableView;
@property(nonatomic)NSArray *storeArray;
@end

@implementation StoreViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self fetchStoreData];
}


-(void)viewWillAppear:(BOOL)animated
{
    self.title = [self.productDictionary objectForKey:@"name"];
}



-(void)fetchStoreData
{
    self.tableView.hidden = YES;

    NSURLSession *session = [NSURLSession sharedSession];
    NSURLSessionDataTask *task = [session dataTaskWithURL:[NSURL URLWithString:[NSString stringWithFormat:@"http://ontariobeerapi.ca:80/products/%@/stores/",[self.productDictionary objectForKey:@"product_id"]]] completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
        
        self.storeArray = [NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:nil];
        
        if(error != nil)
        {
            UIAlertView *alertView = [[UIAlertView alloc]initWithTitle:@"Error" message:error.description delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil];
            [alertView show];
            self.tableView.hidden = YES;

        }
        else if([self.storeArray count] == 0)
        {
            UIAlertView *alertView = [[UIAlertView alloc]initWithTitle:@"No stores has this product" message:@"" delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil];
            [alertView show];
            self.tableView.hidden = YES;
                                      
        }
        else
        {
            self.tableView.hidden = NO;
        }
   
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.tableView reloadData];
            
        });
        
        
    }];
    [task resume];
    
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


- (NSInteger)tableView:(UITableView *)tableView
 numberOfRowsInSection:(NSInteger)section
{
    return [self.storeArray count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView
         cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    StoreTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"StoreTableViewCell"];
    NSDictionary *beerDictionary = [self.storeArray objectAtIndex:indexPath.row];
    cell.titleLabel.text = [beerDictionary objectForKey:@"name"];
    cell.cityLabel.text = [beerDictionary objectForKey:@"city"];
    return cell;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return 80.0;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    MapViewController *mapViewController = [self.storyboard instantiateViewControllerWithIdentifier:@"MapViewController"];
    mapViewController.storeDictionary = [self.storeArray objectAtIndex:indexPath.row];
    [self.navigationController pushViewController:mapViewController animated:YES];
    self.title = @"";

}


@end

File: Swift/BeerSampleSwift/BeerSampleSwift/StoreViewController.swift
//
//  StoreViewController.swift
//  BeerSampleSwift
//
//  Created by Vinita Miranda on 5/25/15.
//  Copyright (c) 2015 Vinita Miranda. All rights reserved.
//

import UIKit

class StoreViewController: UIViewController,DataManagerProtocol,UITableViewDelegate,UITableViewDataSource   {

    @IBOutlet weak var tableView: UITableView!
    var productDictionary : NSDictionary?
    var storeArray = NSMutableArray()
    override func viewDidLoad() {
        super.viewDidLoad()
        self.tableView.registerNib(UINib(nibName: "CustomTableViewCell", bundle: NSBundle.mainBundle()), forCellReuseIdentifier: "CustomTableViewCell")
        self.fetchStoreInfo()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    

    func fetchStoreInfo()
    {
        var dataManager : DataManager = DataManager()
        dataManager.delegate = self
        var storeId  = self.productDictionary?.objectForKey("product_id") as? Int
        dataManager.fetchDataFromUrl("http://ontariobeerapi.ca:80/products/\(storeId!)/stores/")
    }
    
    
    func datafetchCompleted(result : NSMutableArray)
    {
        self.storeArray = result
        dispatch_async(dispatch_get_main_queue())
            {
                self.tableView.reloadData()
        }
    }
    
    func datafetchFailed(error : NSError)
    {
        
    }
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.storeArray.count
    }
    
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell
    {
        var cell : CustomTableViewCell = tableView.dequeueReusableCellWithIdentifier("CustomTableViewCell") as! CustomTableViewCell
        var storeDictionary: NSDictionary = NSDictionary()
        storeDictionary  = self.storeArray[indexPath.row] as! NSDictionary
        cell.titleLabel.text = storeDictionary["name"] as? String
        cell.typeLabel.text = storeDictionary["city"] as? String
        cell.ratingLabel.hidden = true
        cell.titleLeadingSpace.constant = 0;
        cell.typeLeadingSpace.constant = 0;
        cell.titleLabel.layoutIfNeeded()
        cell.titleLabel.setNeedsLayout()
        
        cell.typeLabel.layoutIfNeeded()
        cell.typeLabel.setNeedsLayout()
        
        return cell
    }
    
    
    func tableView(tableView: UITableView, heightForRowAtIndexPath indexPath: NSIndexPath) -> CGFloat {
        return 80
    }

    func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath)
    {
        var mapViewController = self.storyboard!.instantiateViewControllerWithIdentifier("MapViewController") as? MapViewController
        mapViewController?.storeDictionary = self.storeArray[indexPath.row] as? NSDictionary
        self.navigationController?.pushViewController(mapViewController!, animated: true)
    }


}
